pkg/kv/kvserver/replica_lifecycle_datadriven_test.go

```diff
@@ -0,0 +1,340 @@
+func (tc *testCtx) close() {
+	tc.storage.Close()
+}
+
+// newRangeState constructs a new rangeState for the supplied descriptor.
+func newRangeState(desc roachpb.RangeDescriptor) *rangeState {
+	return &rangeState{
+		desc:       desc,
+		version:    cluster.MakeTestingClusterSettings().Version.LatestVersion(),
+		nextReplID: len(desc.InternalReplicas) + 1,
+	}
+}
+
+// mustGetRangeState returns the range state for the given range ID.
+func (tc *testCtx) mustGetRangeState(t *testing.T, rangeID int) *rangeState {
+	rs, ok := tc.ranges[rangeID]
+	if !ok {
+		t.Fatalf("range-id %d not found", rangeID)
+	}
+	return rs
+}
+
+func (tc *testCtx) updatePostReplicaCreateState(
+	t *testing.T, ctx context.Context, rs *rangeState, batch storage.Batch,
+) {
+	// Sanity check that we're not overwriting an existing replica.
+	require.Nil(t, rs.replica)
+	sl := stateloader.Make(rs.desc.RangeID)
+	hs, err := sl.LoadHardState(ctx, batch)
+	require.NoError(t, err)
+	ts, err := sl.LoadRaftTruncatedState(ctx, batch)
+	require.NoError(t, err)
+	replID, err := sl.LoadRaftReplicaID(ctx, batch)
+	require.NoError(t, err)
+	rs.replica = &replicaInfo{
+		id: roachpb.FullReplicaID{
+			RangeID:   rs.desc.RangeID,
+			ReplicaID: replID.ReplicaID,
+		},
+		hs: hs,
+		ts: ts,
+	}
+}
+
+func (rs *rangeState) getReplicaDescriptor() *roachpb.ReplicaDescriptor {
+	for i, repl := range rs.desc.InternalReplicas {
+		if repl.NodeID == roachpb.NodeID(1) {
+			return &rs.desc.InternalReplicas[i]
+		}
+	}
+	panic("replica not found")
+}
+
+func (rs *rangeState) String() string {
+	var sb strings.Builder
+	sb.WriteString(fmt.Sprintf("range %d: [%s,%s) replicas=",
+		rs.desc.RangeID, rs.desc.StartKey, rs.desc.EndKey))
+	for i, replica := range rs.desc.InternalReplicas {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(fmt.Sprintf("n%d(r%d)", replica.NodeID, replica.ReplicaID))
+	}
+	if rs.replica != nil {
+		sb.WriteString(fmt.Sprintf("\n		replica (n1/s1): %s\n", rs.replica))
+	}
+	return sb.String()
+}
+
+func (r *replicaInfo) String() string {
+	var sb strings.Builder
+	sb.WriteString(fmt.Sprintf("id=%s ", r.id))
+	if r.hs == (raftpb.HardState{}) {
+		sb.WriteString("uninitialized")
+	} else {
+		sb.WriteString(fmt.Sprintf("HardState={Term:%d,Vote:%d,Commit:%d} ", r.hs.Term, r.hs.Vote, r.hs.Commit))
+		sb.WriteString(fmt.Sprintf("TruncatedState={Index:%d,Term:%d}", r.ts.Index, r.ts.Term))
+	}
+	return sb.String()
+}
+
+func parseReplicas(t *testing.T, val string) []int {
+	var replicaNodeIDs []int
+	if len(val) >= 2 && val[0] == '[' && val[len(val)-1] == ']' {
+		val = val[1 : len(val)-1]
+		if val != "" {
+			for _, s := range strings.Split(val, ",") {
+				var id int
+				fmt.Sscanf(strings.TrimSpace(s), "%d", &id)
+				replicaNodeIDs = append(replicaNodeIDs, id)
+			}
+		}
+	}
+	// The test is written from the perspective of n1/s1, so not having n1 in
+	// this list should return an error.
+	if !slices.Contains(replicaNodeIDs, 1) {
```

Comment:
`require.Contains`? Or `require.True(t, slices.Contains(...))` if that one doesn't work.